File: js-src/src/storage/repository/InMemoryVcsAdapter.ts
import { Commit } from "../version-control/Commit";
import { CommitGraph } from "../version-control/CommitGraph";
import { getLogger } from "../../logging";
import { VcsAdapter, InternalRepoUpdate } from "./VcsAdapter";

const log = getLogger('InMemoryRepository')


export class InMemoryVcsAdapter implements VcsAdapter {
    private _commitGraph: CommitGraph = new CommitGraph();
    private _commitById: Map<string, Commit> = new Map();

    async getCommitGraph(): Promise<CommitGraph> {
        return CommitGraph.fromData(this._commitGraph.data());
    }

    async getCommits(ids: string[]): Promise<Commit[]> {
        return ids.map((id) => {
            const commit = this._commitById.get(id);
            if (!commit) {
                throw new Error(`Commit ${id} not found`);
            }
            return commit;
        });
    }

    async applyUpdate(update: InternalRepoUpdate): Promise<void> {
        const {
            addedCommits,
            removedCommits,
            updatedCommits,
            addedBranches,
            updatedBranches,
            removedBranches
        } = update;

        // Remove old commits
        for (const commit of removedCommits) {
            this._commitById.delete(commit.id);
            this._commitGraph.removeCommit(commit.id);
        }

        // Update existing commits (full replace: metadata + delta)
        for (const commit of updatedCommits) {
            // Replace full commit in map
            this._commitById.delete(commit.id);
            this._commitById.set(commit.id, commit);
            // Refresh metadata in graph (remove then add)
            this._commitGraph.removeCommit(commit.id);
            this._commitGraph.addCommit(commit.metadata());
        }

        // Add new commits
        for (const commit of addedCommits) {
            this._commitById.set(commit.id, commit);
            this._commitGraph.addCommit(commit.metadata());
        }

        // Add new branches
        for (const branch of addedBranches) {
            this._commitGraph.createBranch(branch.name);
            if (branch.head_commit_id) {
                this._commitGraph.setBranch(branch.name, branch.head_commit_id);
            }
        }

        // Update branches
        for (const branch of updatedBranches) {
            this._commitGraph.setBranch(branch.name, branch.head_commit_id);
        }

        // Remove branches
        for (const branch of removedBranches) {
            this._commitGraph.removeBranch(branch.name);
        }
    }

    close(): void {
        // Nothing to do for in-memory storage
    }

    async eraseStorage(): Promise<void> {
        this._commitGraph = new CommitGraph();
        this._commitById = new Map();
    }
}

File: py-src/fusion/storage/in_memory_vcs_adapter.py
from __future__ import annotations

from fusion import get_logger
from fusion.storage.commit import Commit
from fusion.storage.commit_graph import CommitGraph
from fusion.storage.vcs_adapter import InternalRepoUpdate, VcsAdapter

log = get_logger(__name__)


class InMemoryVcsAdapter(VcsAdapter):
    def __init__(self) -> None:
        self._commit_graph = CommitGraph()
        self._commit_by_id: dict[str, Commit] = {}

    def get_commit_graph(self) -> CommitGraph:
        return CommitGraph.from_data(self._commit_graph.data())

    def get_commits(self, ids: list[str]) -> list[Commit]:
        result: list[Commit] = []
        for commit_id in ids:
            commit = self._commit_by_id.get(commit_id)
            if commit is None:
                raise ValueError(f"Commit {commit_id} not found")
            result.append(commit)
        return result

    def apply_update(self, update: InternalRepoUpdate) -> None:
        # Remove old commits
        for commit in update.removed_commits:
            self._commit_by_id.pop(commit.id, None)
            self._commit_graph.remove_commit(commit.id)

        # Update existing commits (full replace: metadata + delta)
        for commit in update.updated_commits:
            self._commit_by_id.pop(commit.id, None)
            self._commit_by_id[commit.id] = commit
            self._commit_graph.remove_commit(commit.id)
            self._commit_graph.add_commit(commit.metadata())

        # Add new commits
        for commit in update.added_commits:
            self._commit_by_id[commit.id] = commit
            self._commit_graph.add_commit(commit.metadata())

        # Add new branches
        for branch in update.added_branches:
            self._commit_graph.create_branch(branch.name)
            if branch.head_commit_id:
                self._commit_graph.set_branch(branch.name, branch.head_commit_id)

        # Update branches
        for branch in update.updated_branches:
            self._commit_graph.set_branch(branch.name, branch.head_commit_id)

        # Remove branches
        for branch in update.removed_branches:
            self._commit_graph.remove_branch(branch.name)

    def close(self) -> None:
        pass

    def erase_storage(self) -> None:
        self._commit_graph = CommitGraph()
        self._commit_by_id.clear()
